packages/grid/_modules_/grid/hooks/features/pagination/useGridPage.ts

```diff
@@ -150,9 +158,23 @@ export const useGridPage = (
   /**
    * EFFECTS
    */
+  React.useEffect(() => {
+    if (process.env.NODE_ENV !== 'production') {
+      if (props.paginationMode === 'server' && props.rowCount == null) {
+        noRowCountInServerMode();
+      }
+    }
+  }, [props.rowCount, props.paginationMode]);
+
   React.useEffect(() => {
     apiRef.current.setState((state) => {
-      const rowCount = props.rowCount !== undefined ? props.rowCount : visibleTopLevelRowCount;
+      let rowCount;
+      if (props.paginationMode === 'server') {
```

Comment:
Would the following be clearer ?

```suggestion
      if (props.rowCount !== undefined) {
        rowCount = props.rowCount
      } else if (props.paginationMode === 'server') {
        rowCount = state.pagination.rowCount
      } else {
        rowCount = visibleTopLevelRowCount
      }
```